File: python/05.py
from common import *

with open(infile('05.txt')) as f:
    nums = [get_all_ints(line) for line in f]

x_max = max(max(a, b) for a, _, b, _ in nums)
y_max = max(max(c, d) for _, c, _, d in nums)

def sign(x):
    return 0 if x == 0 else x // abs(x)

def dump(grid):
    for row in grid:
        for x in row:
            print(x if x > 0 else '.', end='')
        print()

grid1 = [[0] * (x_max + 1) for _ in range(y_max + 1)]
grid2 = [[0] * (x_max + 1) for _ in range(y_max + 1)]
for x1, y1, x2, y2 in nums:
    dx = x2 - x1
    dy = y2 - y1
    for i in range(max(abs(dx), abs(dy))+1):
        x = x1 + sign(dx) * i
        y = y1 + sign(dy) * i
        if (dx == 0 or dy == 0):
            grid1[y][x] += 1
        grid2[y][x] += 1

p1 = sum(sum(x > 1 for x in row) for row in grid1)
p2 = sum(sum(x > 1 for x in row) for row in grid2)
print("Part 1:", p1)
print("Part 2:", p2)

File: c/05.c
#include "common.h"

#define BOARD_SIZE 1024
int board_1[BOARD_SIZE][BOARD_SIZE] = {0};
int board_2[BOARD_SIZE][BOARD_SIZE] = {0};

int board_result(int board[BOARD_SIZE][BOARD_SIZE]) {
  int count = 0;
  for (int i = 0; i < BOARD_SIZE; i++)
    for (int j = 0; j < BOARD_SIZE; j++)
      if (board[i][j] > 1)
        count++;
  return count;
}

void dump_board(int mx) {
  for (int i = 0; i < mx; i++) {
    for (int j = 0; j < mx; j++) {
      printf("%d", board_1[i][j]);
    }
    printf("\n");
  }
}

void solve() {
  FILE *f = fopen(INFILE("05.txt"), "r");
  char buf[2048];
  int x1, y1, x2, y2;

  while (fgets(buf, 2048, f)) {
    sscanf(buf, "%d,%d -> %d,%d", &x1, &y1, &x2, &y2);
    int dx = x2 - x1, dy = y2 - y1;
    for (int i = 0; i <= max(abs(dx), abs(dy)); i++) {
      int x = x1 + sign(dx) * i;
      int y = y1 + sign(dy) * i;
      if (x1 == x2 || y1 == y2)
        board_1[y][x] += 1;
      board_2[y][x] += 1;
    }
  }
  printf("Part 1: %d\n", board_result(board_1));
  printf("Part 2: %d\n", board_result(board_2));
  fclose(f);
}

int main() {
  solve();
  return 0;
}